include/nbl/builtin/hlsl/atomics.hlsl

```diff
@@ -0,0 +1,87 @@
+// Copyright (C) 2023 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+#ifndef _NBL_BUILTIN_HLSL_ATOMICS_INCLUDED_
+#define _NBL_BUILTIN_HLSL_ATOMICS_INCLUDED_
+
+#ifndef SHARED_MEM
+#error "Atomics need SHARED_MEM defined to point to uint groupshared array"
+#else
+namespace nbl
+{
+namespace hlsl
+{
+namespace atomics
+{
+	// REVIEW:  How should we handle shared memory access in here? Right now we assume SHARED_MEM is defined.
+	// 			Additionally, is there a point in having templated type or should it be uint and the called 
+	//			uses `asuint()` on different types before calling atomics? Note that Interlocked* API only works 
+	//			on uint and int
+	template<typename T>
+	T atomicAdd(in uint ix, T data)
+	{
+		T orig;
+		InterlockedAdd(SHARED_MEM[ix], data, orig);
```

Comment:
this isn't helping, either force us to use `InterlockedXXX` whenever we want a `atomicXXX` 
For a robust solution to this we need to wait for HLSL202x to implement references (its already in the works) or lambdas
that does `#define atomicAdd(VAR,DATA) {T orig; InterlockedAdd(VAR,DATA,orig); return orig;}`

Also stuff like this should be in a `/builtin/hlsl/glsl_compat.hlsl` header

Reply:
Btw is there any limitation that would prevent you from implementing `atomicAdd` as an "inline SPIR-V" ?